openupgradelib/migrator.py

```diff
@@ -0,0 +1,549 @@
+        return allow_pgcodes(self.env.cr, *codes)
+
+    def _execute(self, query, params=None, log_exceptions=True):
+        """Wrapper that logs before executing.
+
+        Parameters are the same as for :meth:`~.Cursor.execute`, except that
+        :param:`log_exceptions` defaults to ``True``.
+        """
+        _logger.info("Executing query with params %s:\n%s", params, query)
+        self.env.cr.execute(query, params, log_exceptions)
+
+    def _fetch_ids(self):
+        """Return list of ids from last executed query."""
+        return tuple(row["id"] for row in self.env.cr.dictfetchall())
+
+    def _model_id(self, model_name):
+        """Get a model's ID in ``ir.model`` table.
+
+        :param str model_name:
+            The model name formatted like ``res.partner``.
+        """
+        self._execute(
+            "SELECT id FROM ir_model WHERE model = %s",
+            (model_name,))
+        return self.env.cr.fetchone()[0]
+
+    def _record_remove(self, model_name, ids, xmlid=True):
+        """Remove a record from the database.
+
+        :param str model_name:
+            Model name from where to remove the record. E.g. ``res.partner``.
+
+        :param int ids:
+            Tuple of IDs of the record to remove.
+
+        :param bool xmlid:
+            Wether to remove the XMLID for this record in this addon too.
+        """
+        table = self._table_name(model_name)
+        self._execute("DELETE FROM %s WHERE id IN %%s" % table, (ids,))
+        if xmlid:
+            self._execute(
+                """ DELETE FROM ir_model_data
+                    WHERE module = %s AND model = %s AND res_id IN %s
+                """,
+                (self.addon, model_name, ids))
+
+    @wip
+    def _table_fk(self, table_name):
+        """Get a list of FK constraints pointing to a given DB table.
+
+        .. warning::
+            The constraints are found in other tables, but point to
+            :param:`table_name`.
+
+        :param str table_name:
+            Table name for whom to get the constraints. E.g. ``res_partner``.
+        """
+        # See http://stackoverflow.com/a/1152321/1468388
+        self._execute(
+            """ SELECT
+                    tc.*,
+                    kcu.column_name,
+                    ccu.table_name AS foreign_table_name,
+                    ccu.column_name AS foreign_column_name
+                FROM
+                    information_schema.table_constraints AS tc
+
+                    JOIN information_schema.key_column_usage AS kcu
+                    ON tc.constraint_name = kcu.constraint_name
+
+                    JOIN information_schema.constraint_column_usage AS ccu
+                    ON tc.constraint_name = ccu.constraint_name
+                WHERE
+                    kcu.constraint_type = 'FOREIGN KEY' AND
+                    ccu.table_name = %s
+            """,
+            (table_name,)
+        )
+        return self.env.cr.dictfetchall()
```

Comment:
I think @StefanRijnhart can best estimate where such a method would be useful. Integrating it would then be rather easy I guess. The difficulty here probably was the query construction itself.